### PR TITLE
add View->AutoLayout->Implementation to mdl

### PR DESCRIPTION
**Pull request by beffge**
This PR will add the implementation attribute used by structuring to define how the view is automatically rendered.

### CODE REVIEW

**Comment by beffge** [comment]
Since I only use the structs, I'm not really sure if I missed anything relevant. Please feel free to point me to any missing parts.

**Comment by raphael** [comment]
Looks great! I would grep the source for where `RankDirection` is used as the usage pattern is going to be pretty much identical. In particular the DSL and expressions also need to be updated as well as [modelizeProps](https://github.com/goadesign/model/blob/bd356bab5b012a932369d956706a7ec8e6e95915/mdl/eval.go#L265).

**Comment by beffge** [comment]
I think I might have found it all 🤞

**Comment by raphael** [comment]
Looks like you got them all! However I don't think we can make a breaking change to the DSL, how about we introduce a new `Implementation` DSL keyword that should be used within a `AutoLayout` expression instead?

**Comment by beffge** [comment]
I warned you, I have no clue about all "other parts" than the c4 structs 😜  But ok, valid point. Let's see if I understood you right before I change it all. It the following example how you envisioned it?

```go
            AutoLayout(RankTopBottom, func() {
                Implementation(ImplementationGraphviz)
                RankSeparation(300)
                NodeSeparation(600)
                EdgeSeparation(200)
                RenderVertices()
            })
```

**Comment by raphael** [comment]
Yup, exactly!

**Comment by beffge** [comment]
Done 🤞

File: dsl/views.go [hunk 1]
```diff
@@ -1838,6 +1840,36 @@ func Position(pos int) {
...
+//	    Views(func() {
+//	        SystemContextView(SoftwareSystem, "context", "An overview diagram.", func() {
+//	            AutoLayout(func() {
+//	                Implementation(ImplementationGraphviz)
+//	            })
+//	        })
+//	    })
+//	})
+func Implementation(impl ImplementationKind) {
+	if reflect.ValueOf(impl).IsNil() {
```
Comment (raphael) [inline dsl/views.go:1862]: Not sure I understand why this test is needed? Isn't ImplementationKind a int? I believe this check will cause a panic.

**Review by raphael — COMMENTED** [review]
Thank you, looks great! Left one comment that would be good to address before merging (should just be removing a few lines).

File: dsl/views.go [hunk 2]
```diff
@@ -1488,6 +1489,7 @@ var DefaultEdgeSeparation = 200
 //	        SystemContextView(SoftwareSystem, "context", "An overview diagram.", func() {
 //	            AddDefault()
 //	            AutoLayout(RankTopBottom, ImplementationDagre, func() {
+//					Implementation(ImplementationGraphviz)
```
[the remark this answers is not in the document]
Comment (beffge) [reply dsl/views.go]: fixed

File: dsl/views.go (hunk 1 above)
Comment (beffge) [reply dsl/views.go]: Very valid, since you need to provide any ImplementationKind and can't just leave it empty. ➡️removed

**Comment by beffge** [comment]
Hey, did you have any chance to review the new changes?

**Comment by beffge** [comment]
Hey @raphael, did you have any chance to review the new changes?

File: dsl/views.go [hunk 3]
```diff
@@ -1451,37 +1464,39 @@ var DefaultNodeSeparation = 600
...
+// AutoLayout enables automatic layout mode for the diagram.
+//
+// The First argument indicates the rank direction, it must be one of
 // RankTopBottom, RankBottomTop, RankLeftRight or RankRightLeft
 //
 // AutoLayout must appear in SystemLandscapeView, SystemContextView,
 // ContainerView, ComponentView, DynamicView or DeploymentView.
 //
-// AutoLayout accepts one or two arguments: the layout rank direction and
+// AutoLayout accepts two or three arguments: the layout rank direction and
```
Comment (raphael) [inline dsl/views.go:1475]: I believe this needs to be changed back

File: dsl/views.go [hunk 4]
```diff
@@ -1451,37 +1464,39 @@ var DefaultNodeSeparation = 600
...
-//	            AutoLayout(RankTopBottom, func() {
-//	                RankSeparation(200)
-//	                NodeSeparation(100)
-//	                EdgeSeparation(10)
-//	                Vertices()
-//	            })
-//	        })
-//	    })
-//	})
+//		var _ = Design(func() {
```
Comment (raphael) [inline dsl/views.go:1480]: The indentation seems a bit wacky :) We should probably just keep the original indent depth and simply add the new line.

File: dsl/views.go [hunk 5]
```diff
@@ -1491,15 +1506,16 @@ func AutoLayout(rank RankDirectionKind, args ...func()) {
 	var dsl func()
 	if len(args) > 0 {
 		dsl = args[0]
-		if len(args) > 1 {
```
Comment (raphael) [inline dsl/views.go:1494]: This should be changed back

**Review by raphael — COMMENTED** [review]
Thank you for the changes, the new function looks great however some of the old changes need to be undone.

**Comment by raphael** [comment]
> Hey @raphael, did you have any chance to review the new changes?

Thanks for the ping, I just gave it another pass, looks great, I believe it just needs a few more tweaks

File: dsl/views.go (hunk 4 above)
Comment (beffge) [reply dsl/views.go:1480]: 🙈 I think it's the auto formatting of my IDE. Took a while to get it back to not change the whole thing 😰

**Comment by beffge** [comment]
Sorry for the missed changes I did not revert. I think we should be good now.

**Comment by raphael** [comment]
Perfect, thank you!

**Comment by beffge** [comment]
gr8, thanks! P.S: Nice new profile pic 💪

**Comment by raphael** [comment]
Haha thank you! I just released v1.9.1 with this addition - thanks again for the contribution!

**Pull request merged**